### PR TITLE
R2RDump - Build integration

### PR DESCRIPTION
**Pull request by acmyu**

### CODE REVIEW

**Comment by A-And** [comment]
@nattress - I might have steered Amy the wrong way with not building r2rdump with the rest of CoreCLR.

@acmyu - `build-managed.cmd` is a wrapper over `run.cmd` in corefx exactly like e.g. `build-packages.cmd` is in coreclr. It defines a build-managed command in its config.json, which invokes msbuild to build .NET Core projects. Bringing over a similar infrastructure would definitely be useful.

**Comment by jkotas** [comment]
Also, check with https://github.com/dotnet/source-build folks on how to wire this in a way compatible with build from source. The mainstream CoreCLR build cannot reference live CoreFX bits because of it would create a cycle.

**Comment by jkotas** [comment]
I am not expert here.

@eerhardt @dagood Any source build concerns wrt building this managed tool in CoreCLR repo?

**Comment by dagood** [comment]
> Any source build concerns wrt building this managed tool in CoreCLR repo?

As long as building it can easily be disabled for source-build and disabling it doesn't change the .NET Core Runtime/SDK end product, there is no source-build concern. If the plan is to bundle this tool in the Runtime or SDK, then there needs to be some analysis of what Source-Build would be missing out on.

If we end up needing to build it in Source-Build, the first thing I notice are the package dependencies in [R2RDump.csproj](https://github.com/acmyu/coreclr/blob/bae564e7ef89a21eb5dddd7c9290f037b28b2760/src/tools/r2rdump/R2RDump.csproj).

 * Microsoft.NETCore.CoreDisTools: I'm unfamiliar with this package.
 * System.CommandLine: This is a CoreFXLab package, and CoreFXLab isn't source-built.
 * System.Reflection.Metadata: This is a CoreFX dependency, but yeah, I imagine you could use the version from the previous release so being circular isn't necessarily a blocker.

/cc @dleeapho

**Comment by eerhardt** [comment]
> If the plan is to bundle this tool in the Runtime or SDK

Can someone point to some documentation (or write it here) on what the expected usage and shipping vehicle is? Is it for internal purposes, or are we expecting it to be shipped as part of our official product?

> If we end up needing to build it in Source-Build, the first thing I notice are the package dependencies

Agreed, this will need to be fixed if this is an official product.

For sure it will need to remove its dependency on `System.CommandLine`. From my understanding that library has no future.  (The easiest way to remove the dependency is to copy the code into the tool's .csproj.)

As for the "live" corefx dependencies, you won't be able to do that. Like @dagood says above, you can compile against previously shipped versions - this is what repos like dotnet/roslyn do. They compile against older versions of the `System` packages, like `System.Runtime v4.3.0`.  Is it possible to use the previously shipped `System.Reflection.Metadata v1.6.0`? If not, then the tool would have to move "higher" in the repo hierarchy.

And finally `Microsoft.NETCore.CoreDisTools` isn't open-source code, is it?  If not, than that will definitely be a blocker for source-build.

**Comment by acmyu** [comment]
@dagood @eerhardt Microsoft.NETCore.CoreDisTools's source code is in the LLILC repo. Would that be okay for building this?

Kathleen Dollard mentioned that there will be a replacement for System.Commandline soon, but I should use System.Commandline for now. By copying the code to the tool's csproj, do you mean having a copy of System.Commandline in R2RDump and referencing that instead of the one in CorFXLab?

**Comment by eerhardt** [comment]
OK, so since this tool is not part of the official product, then we can just disable building it during source-build.  To do that, there is a special parameter that is passed during source-build `DotNetBuildFromSource=true`.  You can check for this during the build and skip building this tool when that parameter is set.

Here are a few examples from other repos:

https://github.com/dotnet/corefx/blob/7be44d2de8661cd925955e540d940b4eb47e5753/external/dir.proj#L15
https://github.com/dotnet/roslyn/blob/3a92df3fa2b80b5ef603962114c912a7230d57d8/build/Targets/Imports.targets#L113
https://github.com/dotnet/standard/blob/ad9c483514d3fe85215d963cc5a9a70865249fcb/src.builds#L8-L11

File: build.cmd [hunk 1]
```diff
@@ -326,6 +326,9 @@ REM ============================================================================
 
 @if defined _echo @echo on
 
+@call dotnet restore %__ProjectDir%\src\tools\r2rdump
```
Comment (eerhardt) [inline build.cmd]: You probably shouldn't even restore the project during source-build.

**Comment by acmyu** [comment]
@dotnet-bot Test this please

File: src/build.proj [hunk 2]
```diff
@@ -5,6 +5,7 @@
   <ItemGroup>
     <Project Condition="$(SkipSOS) != 'true'" Include="ToolBox\SOS\NETCore\SOS.NETCore.csproj" />
     <Project Include="System.Private.CoreLib\System.Private.CoreLib.csproj" />
+    <Project Condition="'$(DotNetBuildFromSource)' != 'true' AND '$(__BuildManagedTools)' == '1'" Include="tools/r2rdump/R2RDump.csproj" />
```
Comment (zamont) [inline src/build.proj]: In build.cmd, the `%__BuildSOS%` environment variable is used to set `$(SkipSOS)` to true. You could do the same thing here with the `%__BuildManagedTools%` variable so here you could test whether or not `$(BuildManagedTools)` is true.

**Comment by zamont** [comment]
Looks good to me! Once we resolve the CI issues, I think this is good to go!

**Comment by zamont** [comment]
Can you add to the Clean target in build.proj to remove any files from the R2RDump build? https://github.com/dotnet/coreclr/blob/13f2f8ce261db0771f6b602e90f0a408cdd33142/build.proj#L23

File: src/tools/r2rdump/R2RDump.csproj [hunk 3]
```diff
@@ -1,18 +1,29 @@
...
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
 
   <PropertyGroup>
+    <AssemblyName>R2RDump</AssemblyName>
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <OutputType>Exe</OutputType>
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
+    <NoStdLib>true</NoStdLib>
```
Comment (zamont) [inline src/tools/r2rdump/R2RDump.csproj]: Can you check if we really need to set `<NoStdLib>`, `<NoCompilerStandardLib>`, and `<ExcludeMscorlibFacade>`?

**Comment by zamont** [comment]
@dotnet-bot test this please

**Comment by acmyu** [comment]
@dotnet-bot test this please

**Comment by acmyu** [comment]
@dotnet-bot test this please

**Comment by acmyu** [comment]
@A-And The System.Drawing.Common.Tests in the CoreFX Tests is failing

**Comment by A-And** [comment]
@acmyu - Yep, it seems to have regressed after the Jenkins outage. Any chance you could open a separate PR to exclude it?

**Pull request merged**